File: src/git/git.ts
import { spawnSync } from 'node:child_process';
import { existsSync } from 'node:fs';
import { join } from 'node:path';

export interface GitRunResult {
  ok: boolean;
  stdout: string;
  stderr: string;
  code: number | null;
}

export function runGit(root: string, args: string[]): GitRunResult {
  const res = spawnSync('git', args, {
    cwd: root,
    encoding: 'utf8',
    windowsHide: true,
    maxBuffer: 16 * 1024 * 1024,
  });
  return {
    ok: res.status === 0,
    stdout: (res.stdout ?? '').toString(),
    stderr: (res.stderr ?? '').toString(),
    code: res.status,
  };
}

export function isGitRepo(root: string): boolean {
  if (existsSync(join(root, '.git'))) return true;
  const res = runGit(root, ['rev-parse', '--is-inside-work-tree']);
  return res.ok && res.stdout.trim() === 'true';
}

export function currentBranch(root: string): string | null {
  const res = runGit(root, ['branch', '--show-current']);
  if (res.ok && res.stdout.trim()) return res.stdout.trim();
  return null;
}

export function statusShort(root: string): string {
  const res = runGit(root, ['status', '--short']);
  return res.ok ? res.stdout.trimEnd() : `(git unavailable: ${res.stderr.trim()})`;
}

export function diffStat(root: string): string {
  const res = runGit(root, ['diff', '--stat']);
  return res.ok ? res.stdout.trimEnd() : '';
}

export function diffUnified(root: string, maxLines = 300): string {
  const res = runGit(root, ['diff']);
  if (!res.ok) return '';
  const lines = res.stdout.split('\n');
  if (lines.length <= maxLines) return res.stdout;
  return lines.slice(0, maxLines).join('\n') + `\n… [diff truncated, ${lines.length - maxLines} more lines]`;
}

export function recentLog(root: string, n = 5): string {
  const res = runGit(root, ['log', `-${n}`, '--oneline', '--no-decorate']);
  return res.ok ? res.stdout.trimEnd() : '';
}

/** Compact git context injected into the agent's system prompt. */
export function gitAwareness(root: string): string {
  const g = gitSummary(root);
  if (!g.isRepo) return '(not a git repository)';
  const parts = [`branch: ${g.branch ?? 'detached'}`, `working tree: ${g.hasChanges ? `${g.statusLines} change(s)` : 'clean'}`];
  const status = statusShort(root);
  if (status.trim()) parts.push(`status:\n${status.split('\n').slice(0, 15).join('\n')}`);
  const recent = recentLog(root, 3);
  if (recent.trim()) parts.push(`recent commits:\n${recent}`);
  return parts.join('\n');
}

export interface GitSummary {
  branch: string | null;
  statusLines: number;
  hasChanges: boolean;
  isRepo: boolean;
}

export function gitSummary(root: string): GitSummary {
  const isRepo = isGitRepo(root);
  if (!isRepo) return { branch: null, statusLines: 0, hasChanges: false, isRepo: false };
  const short = statusShort(root);
  const statusLines = short.split('\n').filter((l) => l.trim().length > 0).length;
  return { branch: currentBranch(root), statusLines, hasChanges: statusLines > 0, isRepo: true };
}

File: grace/git.py
"""Git helpers (port of src/git/git.ts)."""

import os
import subprocess


def run_git(root: str, args: list[str]) -> dict:
    try:
        res = subprocess.run(
            ["git", *args],
            cwd=root,
            capture_output=True,
            text=True,
            encoding="utf-8",
            errors="replace",
            timeout=30,
        )
        return {"ok": res.returncode == 0, "stdout": res.stdout or "", "stderr": res.stderr or "", "code": res.returncode}
    except Exception:
        return {"ok": False, "stdout": "", "stderr": "", "code": None}


def is_git_repo(root: str) -> bool:
    if os.path.exists(os.path.join(root, ".git")):
        return True
    res = run_git(root, ["rev-parse", "--is-inside-work-tree"])
    return res["ok"] and res["stdout"].strip() == "true"


def current_branch(root: str) -> str | None:
    res = run_git(root, ["branch", "--show-current"])
    if res["ok"] and res["stdout"].strip():
        return res["stdout"].strip()
    return None


def status_short(root: str) -> str:
    res = run_git(root, ["status", "--short"])
    if res["ok"]:
        return res["stdout"].rstrip()
    return f"(git unavailable: {res['stderr'].strip()})"


def diff_stat(root: str) -> str:
    res = run_git(root, ["diff", "--stat"])
    return res["stdout"].rstrip() if res["ok"] else ""


def diff_unified(root: str, max_lines: int = 300) -> str:
    res = run_git(root, ["diff"])
    if not res["ok"]:
        return ""
    lines = res["stdout"].split("\n")
    if len(lines) <= max_lines:
        return res["stdout"]
    return "\n".join(lines[:max_lines]) + f"\n… [diff truncated, {len(lines) - max_lines} more lines]"


def recent_log(root: str, n: int = 5) -> str:
    res = run_git(root, ["log", f"-{n}", "--oneline", "--no-decorate"])
    return res["stdout"].rstrip() if res["ok"] else ""


def git_awareness(root: str) -> str:
    """Compact git context injected into the agent's system prompt."""
    g = git_summary(root)
    if not g["isRepo"]:
        return "(not a git repository)"
    parts = [f"branch: {g['branch'] or 'detached'}", f"working tree: {g['statusLines']} change(s)" if g["hasChanges"] else "working tree: clean"]
    status = status_short(root)
    if status.strip():
        parts.append("status:\n" + "\n".join(status.split("\n")[:15]))
    recent = recent_log(root, 3)
    if recent.strip():
        parts.append("recent commits:\n" + recent)
    return "\n".join(parts)


def git_summary(root: str) -> dict:
    repo = is_git_repo(root)
    if not repo:
        return {"branch": None, "statusLines": 0, "hasChanges": False, "isRepo": False}
    short = status_short(root)
    status_lines = len([l for l in short.split("\n") if l.strip()])
    return {"branch": current_branch(root), "statusLines": status_lines, "hasChanges": status_lines > 0, "isRepo": True}
